### PR TITLE
Deberta v2 support

### PR DESCRIPTION
**Pull request by vrdn-23**
@kozistr thanks again for agreeing to help out!

### CODE REVIEW

File: backends/candle/src/models/debertav2.rs [hunk 1]
```diff
@@ -0,0 +1,1490 @@
...
+pub const DTYPE: DType = DType::F32;
+
+// NOTE: HiddenAct and HiddenActLayer are both direct copies from bert.rs.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum HiddenAct {
+    Gelu,
+    GeluApproximate,
+    Relu,
+}
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: We could use [HiddenAct](https://github.com/huggingface/text-embeddings-inference/blob/6e900afba71821fdf250e380d7da1f5a6e5e7e27/backends/candle/src/layers/linear.rs#L7) enum instead of creating a new one!

File: backends/candle/src/models/debertav2.rs [hunk 2]
```diff
@@ -0,0 +1,1490 @@
...
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct DebertaV2Config {
+    pub vocab_size: usize,
+    pub hidden_size: usize,
+    pub num_hidden_layers: usize,
+    pub num_attention_heads: usize,
+    pub intermediate_size: usize,
+    pub hidden_act: HiddenAct,
+    pub hidden_dropout_prob: f64,
+    pub attention_probs_dropout_prob: f64,
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: We mostly don't use a dropout layer for the inference stage, so maybe we could remove these configs

File: backends/candle/src/models/debertav2.rs [hunk 3]
```diff
@@ -0,0 +1,1490 @@
...
+
+        Ok(output)
+    }
+}
+
+// https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L605
+pub struct DebertaV2DisentangledSelfAttention {
+    config: DebertaV2Config,
+    num_attention_heads: usize,
+    query_proj: candle_nn::Linear,
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: `Linear` is frequently used over the codes, so I guess it'd be better to use `Linear`, rather than `candle_nn::Linear`.

File: backends/candle/src/models/debertav2.rs [hunk 4]
```diff
@@ -0,0 +1,1490 @@
...
+        let t_zeroes = Tensor::new(&[0f32], device)?.broadcast_as(input.shape())?;
+        output = rmask.where_cond(&t_zeroes, &output)?;
+
+        Ok(output)
+    }
+}
+
+// https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L605
+pub struct DebertaV2DisentangledSelfAttention {
+    config: DebertaV2Config,
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: it looks like `config` is not used, i guess including only the necessary configs would be great in terms of compactness.

File: backends/candle/src/models/debertav2.rs [hunk 5]
```diff
@@ -0,0 +1,1490 @@
...
+            None => hidden_states,
+        };
+
+        let query_layer = self.transpose_for_scores(&self.query_proj.forward(query_states)?)?;
+        let key_layer = self.transpose_for_scores(&self.key_proj.forward(query_states)?)?;
+        let value_layer = self.transpose_for_scores(&self.value_proj.forward(query_states)?)?;
+
+        let mut scale_factor: usize = 1;
+
+        if self.config.pos_att_type.iter().any(|s| s == "c2p") {
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: how about creating a variable like `is_c2p_attn` to predefine the pos_att_type instead of searching the attn type on the fly?

or we could do this in the deserializer

File: backends/candle/src/models/debertav2.rs [hunk 6]
```diff
@@ -0,0 +1,1490 @@
...
+            None => candle::bail!("`predict` is not implemented for this model"),
+            Some(classifier) => {
+                let (pooled_embeddings, _raw_embeddings) = self.forward(batch)?;
+                let pooled_embeddings =
+                    pooled_embeddings.expect("pooled_embeddings is empty. This is a bug.");
+                classifier.forward(&pooled_embeddings)
+            }
+        }
+    }
+}
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: EOF

File: backends/candle/src/models/debertav2.rs [hunk 7]
```diff
@@ -0,0 +1,1490 @@
...
+    position_embeddings: Option<Embedding>,
+    token_type_embeddings: Option<Embedding>,
+    layer_norm: LayerNorm,
+    position_ids: Tensor,
+    config: DebertaV2Config,
+    embedding_size: usize,
+    embed_proj: Option<candle_nn::Linear>,
+}
+
+impl DebertaV2Embeddings {
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs:117]: looks like, in TEI, each `impl` has a `span`.

File: backends/candle/src/models/debertav2.rs [hunk 8]
```diff
@@ -0,0 +1,1490 @@
...
+            embedding_size,
+            embed_proj,
+        })
+    }
+
+    pub fn forward(
+        &self,
+        input_ids: Option<&Tensor>,
+        token_type_ids: Option<&Tensor>,
+        position_ids: Option<&Tensor>,
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: We always have their values (`input_ids`, `token-type_ids`, `position_ids`), so I think `Tensor` type would be enough

File: backends/candle/src/models/debertav2.rs [hunk 9]
```diff
@@ -0,0 +1,1490 @@
...
+        }
+
+        Ok(embeddings)
+    }
+}
+
+// https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L72
+struct XSoftmax {}
+
+impl XSoftmax {
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: I guess, we don't need this layer! Just add the attention mask to the input like [this](https://github.com/huggingface/text-embeddings-inference/blob/6e900afba71821fdf250e380d7da1f5a6e5e7e27/backends/candle/src/models/modernbert.rs#L261)

File: backends/candle/src/models/debertav2.rs [hunk 10]
```diff
@@ -0,0 +1,1490 @@
...
+
+        let embedding_output = self.embeddings.forward(
+            Some(&input_ids),
+            Some(&token_type_ids),
+            Some(&position_ids),
+            attention_mask.as_ref(),
+            None,
+        )?;
+
+        let encoder_attention_mask = attention_mask.as_ref().cloned().unwrap_or_else(|| Tensor::ones(shape, DType::I64, &self.device).unwrap());
```
Comment (kozistr) [inline backends/candle/src/models/debertav2.rs]: How about creating the attention mask when batch_size is 1, at line 1202, if we eventually need one?

And it seems like the dtype of the attention mask is float32 when batched input, while int64 in the single input case.

**Review by kozistr — COMMENTED** [review]
great work! I've just left some reviews, and will get back to you when I have some time on my hands :)

File: backends/candle/src/models/debertav2.rs (hunk 1 above)
Comment (vrdn-23) [reply backends/candle/src/models/debertav2.rs]: I added that a custom serializer but unsure if there is a better way to do this than having 2 different functions when the Pooler_Act is optional v/s not.

Any ideas?

**Comment by vrdn-23** [comment]
He @kozistr, I finally got some time to go through this and started addressing your comments. I will mostly raise a PR against the main repo this week, so hoping to get more feedback from you there!

**Comment by vrdn-23** [comment]
I've addressed all the comments in this PR. closing this in favor of https://github.com/huggingface/text-embeddings-inference/pull/746
